File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/EnterprisePhoneFragment.java
package app.editors.manager.ui.fragments.login;

import android.content.Context;
import android.os.Bundle;
import android.telephony.PhoneNumberFormattingTextWatcher;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatEditText;

import moxy.presenter.InjectPresenter;
import com.google.android.material.textfield.TextInputLayout;

import java.util.Locale;

import javax.inject.Inject;

import app.editors.manager.R;
import app.editors.manager.app.App;
import app.editors.manager.managers.tools.CountriesCodesTool;
import app.editors.manager.mvp.presenters.login.EnterprisePhonePresenter;
import app.editors.manager.mvp.views.login.EnterprisePhoneView;
import app.editors.manager.ui.fragments.base.BaseAppFragment;
import app.editors.manager.ui.views.edits.BaseWatcher;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnEditorAction;
import butterknife.Unbinder;


public class EnterprisePhoneFragment extends BaseAppFragment implements EnterprisePhoneView {

    public static final String TAG = EnterprisePhoneFragment.class.getSimpleName();

    private static final String TAG_CODE = "TAG_CODE";
    private static final String TAG_NAME = "TAG_NAME";
    private static final String TAG_REGION = "TAG_REGION";

    protected Unbinder mUnbinder;
    @BindView(R.id.login_phone_country_edit)
    protected AppCompatEditText mLoginPhoneCountryEdit;
    @BindView(R.id.login_phone_country_layout)
    protected TextInputLayout mLoginPhoneCountryLayout;
    @BindView(R.id.login_phone_number_edit)
    protected AppCompatEditText mLoginPhoneNumberEdit;
    @BindView(R.id.login_phone_number_layout)
    protected TextInputLayout mLoginPhoneNumberLayout;
    @BindView(R.id.login_phone_send_button)
    protected AppCompatButton mLoginPhoneSendButton;

    @Inject
    protected CountriesCodesTool mCountriesCodesTool;

    @InjectPresenter
    EnterprisePhonePresenter mEnterprisePhonePresenter;

    private int mCountryCode;
    private String mCountryName;
    private String mCountryRegion;

    public static EnterprisePhoneFragment newInstance() {
        return new EnterprisePhoneFragment();
    }

    public static EnterprisePhoneFragment newInstance(final int code, final String name, final String region) {
        final EnterprisePhoneFragment phoneFragment = new EnterprisePhoneFragment();
        final Bundle bundle = new Bundle();
        bundle.putInt(TAG_CODE, code);
        bundle.putString(TAG_NAME, name);
        bundle.putString(TAG_REGION, region);
        phoneFragment.setArguments(bundle);
        return phoneFragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        App.getApp().getAppComponent().inject(this);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        final View view = inflater.inflate(R.layout.fragment_login_enterprise_phone, container, false);
        mUnbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init(savedInstanceState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
        mEnterprisePhonePresenter.cancelRequest();
    }

    @OnClick(R.id.login_phone_send_button)
    protected void sendSmsClick() {
        final String phoneNumber = mLoginPhoneNumberEdit.getText().toString().trim().replace(" ", "");
        final String validNumber = mCountriesCodesTool.getPhoneE164(phoneNumber, mCountryRegion);
        if (validNumber != null) {
            showWaitingDialog(getString(R.string.dialogs_wait_title));
            mEnterprisePhonePresenter.setPhone(validNumber);
        } else {
            final String message = getString(R.string.login_sms_phone_error_format);
            mLoginPhoneNumberLayout.setError(message);
        }
    }

    @OnEditorAction(R.id.login_phone_number_edit)
    protected boolean actionKeyPress(final TextView v, final int actionId, final KeyEvent event) {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            sendSmsClick();
            return true;
        }

        return false;
    }

    @Override
    public void onError(@Nullable String message) {
        hideDialog();
        showSnackBar(message);
    }

    @Override
    public void onSuccessChange() {
        hideDialog();
        showFragment(EnterpriseSmsFragment.newInstance(false, null), EnterpriseSmsFragment.TAG, false);
    }

    private void init(final Bundle savedInstanceState) {
        setActionBarTitle(getString(R.string.login_sms_phone_number_verification));
        mLoginPhoneCountryEdit.setOnClickListener(v -> showFragment(CountriesCodesFragment.newInstance(), CountriesCodesFragment.TAG, false));
        final CountriesCodesTool.Codes codes = mCountriesCodesTool.getCodeByRegion(Locale.getDefault().getCountry());
        if (codes != null) {
            mCountryCode = codes.mNumber;
            mCountryName = codes.mName;
            mCountryRegion = codes.mCode;
        }

        showKeyboard(mLoginPhoneNumberEdit);
        mLoginPhoneNumberEdit.addTextChangedListener(new PhoneNumberFormattingTextWatcher());
        mLoginPhoneNumberEdit.setText("+" + mCountryCode);
        mLoginPhoneNumberEdit.addTextChangedListener(new FieldsWatcher());

        mLoginPhoneCountryEdit.setText(mCountryName);
        mLoginPhoneCountryEdit.setKeyListener(null);

        final Bundle bundle = getArguments();
        if (bundle != null) {
            if (bundle.containsKey(TAG_CODE) && bundle.containsKey(TAG_NAME) && bundle.containsKey(TAG_REGION)) {
                mCountryCode = bundle.getInt(TAG_CODE);
                mCountryName = bundle.getString(TAG_NAME);
                mCountryRegion = bundle.getString(TAG_REGION);
                mLoginPhoneCountryEdit.setText(mCountryName);
                mLoginPhoneNumberEdit.setText("+" + mCountryCode);
            }
        }

        final int position = mLoginPhoneNumberEdit.getText().toString().length();
        mLoginPhoneNumberEdit.setSelection(position);
    }

    /*
     * Phone edit field
     * */
    private class FieldsWatcher extends BaseWatcher {

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            mLoginPhoneNumberLayout.setErrorEnabled(false);
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/EnterprisePhoneFragment.kt
package app.editors.manager.ui.fragments.login

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.telephony.PhoneNumberFormattingTextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import androidx.fragment.app.viewModels
import app.editors.manager.R
import app.editors.manager.app.appComponent
import app.editors.manager.app.loginComponent
import app.editors.manager.databinding.FragmentLoginEnterprisePhoneBinding
import app.editors.manager.managers.tools.CountriesCodesTool
import app.editors.manager.ui.fragments.base.BaseAppFragment
import app.editors.manager.ui.views.edits.BaseWatcher
import app.editors.manager.viewModels.login.EnterprisePhoneState
import app.editors.manager.viewModels.login.EnterprisePhoneViewModel
import app.editors.manager.viewModels.login.EnterprisePhoneViewModelFactory
import javax.inject.Inject

class EnterprisePhoneFragment : BaseAppFragment() {

    companion object {
        @JvmField
        val TAG: String = EnterprisePhoneFragment::class.java.simpleName

        private const val TAG_CODE = "TAG_CODE"
        private const val TAG_NAME = "TAG_NAME"
        private const val TAG_REGION = "TAG_REGION"
        private const val TAG_REQUEST = "TAG_REQUEST"

        fun newInstance(request: String?): EnterprisePhoneFragment {
            return EnterprisePhoneFragment().apply {
                arguments = Bundle(1).apply {
                    putString(TAG_REQUEST, request)
                }
            }
        }

        fun newInstance(code: Int, name: String?, region: String?): EnterprisePhoneFragment {
            return EnterprisePhoneFragment().apply {
                arguments = Bundle(3).apply {
                    putInt(TAG_CODE, code)
                    putString(TAG_NAME, name)
                    putString(TAG_REGION, region)
                }
            }
        }
    }

    private val viewModel by viewModels<EnterprisePhoneViewModel> {
        EnterprisePhoneViewModelFactory(requireContext().loginComponent.loginService)
    }

    @Inject
    lateinit var countriesCodesTool: CountriesCodesTool

    private var countryCode = 0
    private var countryName: String? = null
    private var countryRegion: String? = null

    private var viewBinding: FragmentLoginEnterprisePhoneBinding? = null

    override fun onDestroyView() {
        super.onDestroyView()
        viewBinding = null
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        requireContext().appComponent.inject(viewModel)
        requireContext().appComponent.inject(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        viewBinding = FragmentLoginEnterprisePhoneBinding.inflate(inflater)
        return viewBinding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    private fun sendSmsClick() {
        val phoneNumber = viewBinding?.loginPhoneNumberEdit?.text.toString().trim { it <= ' ' }.replace(" ", "")
        val validNumber = countriesCodesTool.getPhoneE164(phoneNumber, countryRegion)
        if (validNumber != null) {
            showWaitingDialog(getString(R.string.dialogs_wait_title))
            arguments?.getString(TAG_REQUEST)?.let { viewModel.setPhone(validNumber, it) }
        } else {
            val message = getString(R.string.login_sms_phone_error_format)
            viewBinding?.loginPhoneNumberLayout?.error = message
        }
    }


    private fun actionKeyPress(actionId: Int): Boolean {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            sendSmsClick()
            return true
        }
        return false
    }

    private fun onError(message: String?) {
        hideDialog()
        message?.let { showSnackBar(it) }
    }

    private fun onSuccessChange(request: String) {
        hideDialog()
        showFragment(
            EnterpriseSmsFragment.newInstance(false, request),
            EnterpriseSmsFragment.TAG,
            false
        )
    }

    @SuppressLint("SetTextI18n")
    private fun init() {
        setActionBarTitle(getString(R.string.login_sms_phone_number_verification))
        initListeners()
        viewModel.countiesCodes.observe(viewLifecycleOwner) { codes: CountriesCodesTool.Codes? ->
            codes?.let {
                countryCode = codes.number
                countryName = codes.name
                countryRegion = codes.code
                showKeyboard(viewBinding?.loginPhoneNumberEdit)
                viewBinding?.loginPhoneNumberEdit?.setText("+$countryCode")
                viewBinding?.loginPhoneCountryEdit?.apply {
                    setText(countryName)
                    keyListener = null
                }
            }
        }
//        val codes = countriesCodesTool.getCodeByRegion(Locale.getDefault().country)
//        if (codes != null) {
//            countryCode = codes.number
//            countryName = codes.name
//            countryRegion = codes.code
//        }
//        showKeyboard(viewBinding?.loginPhoneNumberEdit)
//        viewBinding?.loginPhoneNumberEdit?.setText("+$countryCode")
//        viewBinding?.loginPhoneCountryEdit?.apply {
//            setText(countryName)
//            keyListener = null
//        }
//        val bundle = arguments
//        if (bundle != null) {
//            if (bundle.containsKey(TAG_CODE) && bundle.containsKey(TAG_NAME) && bundle.containsKey(
//                    TAG_REGION
//                )
//            ) {
//                countryCode = bundle.getInt(TAG_CODE)
//                countryName = bundle.getString(TAG_NAME)
//                countryRegion = bundle.getString(TAG_REGION)
//                viewBinding?.loginPhoneCountryEdit?.setText(countryName)
//                viewBinding?.loginPhoneNumberEdit?.setText("+$countryCode")
//            }
//        }
        val position = viewBinding?.loginPhoneNumberEdit?.text.toString().length
        viewBinding?.loginPhoneNumberEdit?.setSelection(position)
    }

    private fun initListeners() {
        viewBinding?.loginPhoneSendButton?.setOnClickListener {
            sendSmsClick()
        }

        viewBinding?.loginPhoneNumberEdit?.setOnEditorActionListener { _, actionId, _ ->
            actionKeyPress(actionId)
        }

        viewBinding?.loginPhoneCountryEdit?.setOnClickListener {
            showFragment(
                CountriesCodesFragment.newInstance(),
                CountriesCodesFragment.TAG,
                false
            )
        }

        viewBinding?.loginPhoneNumberEdit?.apply {
            addTextChangedListener(PhoneNumberFormattingTextWatcher())
            addTextChangedListener(FieldsWatcher())
        }

        viewModel.stateLiveData.observe(viewLifecycleOwner) { state: EnterprisePhoneState? ->
            when (state) {
                is EnterprisePhoneState.Success -> {
                    onSuccessChange(state.request)
                }
                is EnterprisePhoneState.Error -> {
                    onError(state.message)
                }
            }
        }
    }

    /*
     * Phone edit field
     * */
    private inner class FieldsWatcher : BaseWatcher() {
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            viewBinding?.loginPhoneNumberLayout?.isErrorEnabled = false
        }
    }
}